sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1157 @@
+
+    keys : list of str
+        A list of strings which are accepted parameters by the created
+        function, e.g. ``["sample_weight"]`` if the corresponding method
+        accepts it as a metadata.
+
+    Notes
+    -----
+    This class is a descriptor [1]_ and uses PEP-362 to set the signature of
+    the returned function [2]_.
+
+    References
+    ----------
+    .. [1] https://docs.python.org/3/howto/descriptor.html
+
+    .. [2] https://www.python.org/dev/peps/pep-0362/
+    """
+
+    def __init__(self, name, keys):
+        self.name = name
+        self.keys = keys
+
+    def __get__(self, instance, owner):
+        # we would want to have a method which accepts only the expected args
+        def func(**kw):
+            if set(kw) - set(self.keys):
+                raise TypeError(
+                    f"Unexpected args: {set(kw) - set(self.keys)}. Accepted arguments"
+                    f" are: {set(self.keys)}"
+                )
+
+            requests = instance._get_metadata_request()
+            method_metadata_request = getattr(requests, self.name)
+
+            for prop, alias in kw.items():
+                if alias is not UNCHANGED:
+                    method_metadata_request.add_request(param=prop, alias=alias)
+            instance._metadata_request = requests
+
+            return instance
+
+        # Now we set the relevant attributes of the function so that it seems
+        # like a normal method to the end user, with known expected arguments.
+        func.__name__ = f"set_{self.name}_request"
+        params = [
+            inspect.Parameter(
+                name="self",
+                kind=inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                annotation=owner,
+            )
+        ]
+        params.extend(
+            [
+                inspect.Parameter(
+                    k,
+                    inspect.Parameter.KEYWORD_ONLY,
+                    default=UNCHANGED,
+                    annotation=Optional[Union[RequestType, str]],
+                )
+                for k in self.keys
+            ]
+        )
+        func.__signature__ = inspect.Signature(
+            params,
+            return_annotation=owner,
+        )
+        doc = REQUESTER_DOC.format(method=self.name)
+        for metadata in self.keys:
+            doc += REQUESTER_DOC_PARAM.format(metadata=metadata, method=self.name)
+        doc += REQUESTER_DOC_RETURN
+        func.__doc__ = doc
+        return func
+
+
+class _MetadataRequester:
+    """Mixin class for adding metadata request functionality.
+
+    .. versionadded:: 1.1
+    """
+
+    def __init_subclass__(cls, **kwargs):
+        """Set the ``set_{method}_request`` methods.
+
+        This uses PEP-487 [1]_ to set the ``set_{method}_request`` methods. It
+        looks for the information available in the set default values which are
+        set using ``__metadata_request__*`` class attributes.
```

Comment:
```suggestion
        looks for the information available in the set default values which are
        set using ``__metadata_request__*`` class attributes, or inferred
        from method signatures.
```

You might want to note in what cases it might be beneficial to explicitly set `__metadata_request__*` rather than using signature sniffing.